blockchain/chain/src/store/chain_store.rs

```diff
@@ -192,6 +196,27 @@ where
         }
         Ok(())
     }
+
+    pub async fn sub_head_changes(&mut self) -> Result<Receiver<HeadChange>, Error> {
+        let mut subscribed_head_change = self.subscribe();
+        let head = self
+            .heaviest_tipset()
+            .ok_or_else(|| Error::Other("Could not get heaviest tipset".to_string()))?;
+        let (sender, reciever) = channel(16);
+        sender.send(HeadChange::Current(head)).await;
+        task::spawn(async move {
+            while let Some(val) = subscribed_head_change.next().await {
+                if !sender.is_empty() {
+                    warn!(
+                        "head change sub is slow, has {:} buffered entries",
+                        sender.len()
+                    )
+                }
+                sender.send(val).await
+            }
+        });
```

Comment:
A task will be spawned for every function call of this, that won't ever exit. This should probably be one task where the receiver is cloned and passed back in this function?

Reply:
I think ashanti is mostly correct here, except for the fact that the loop never stops. You need to stop the loop once the receiver is dropped, i.e. the websocket connection is terminated. I think here, you could use the futures channels to better handle disconnects https://docs.rs/futures/0.3.5/futures/channel/mpsc/index.html. In async-std channels, i think you only detect a disconnection of all the receivers if the channel is full AND all recievers have been dropped.

Reply:
nice! Shouldn't be hard to switch between the two. Thank you!